h5py/tests/test_dataset.py

```diff
@@ -165,6 +166,58 @@ def test_empty_create_via_Empty_class(self):
         self.f.create_dataset('foo', data=h5py.Empty(dtype='f'))
         self.assertTrue(is_empty_dataspace(self.f['foo'].id))
 
+    def test_create_incompatible_data(self):
+        # Shape tuple is incompatible with data
+        with self.assertRaises(ValueError):
+            self.f.create_dataset('bar', shape=4, data= np.arange(3))
+
+
+class TestReadDirectly(BaseDataset):
+
+    """
+        Feature: Read data directly from Dataset into a Numpy array
+    """
+
+    def test_read_direct(self):
+        dset = self.f.create_dataset("dset", (100,), dtype='int64')
+        empty_dset = self.f.create_dataset("edset", dtype='int64')
+        arr = np.zeros((100,))
+        arr2 = np.zeros((200,))
+
+        # read empty dataset
+        with self.assertRaises(TypeError):
+            empty_dset.read_direct(arr, np.s_[0:10], np.s_[50:60])
+
+        dset.read_direct(arr2, np.s_[0:10], np.s_[50:60])
+        self.assertEqual(dset.shape, (100,))
+
+        # Can't broadcast from source shape (100,) to array shape (200,)
+        with self.assertRaises(TypeError):
+            dset.read_direct(arr2)
+
+class TestWriteDirectly(BaseDataset):
+
+    """
+        Feature: Write Numpy array directly into Dataset
+    """
+
+    def test_write_direct(self):
+        dset = self.f.create_dataset('dset', (100,), dtype='int32')
+        empty_dset = self.f.create_dataset("edset", dtype='int64')
+        arr = np.ones((100,))
+        arr2 = np.zeros((200,))
+
+        # write into empty dataset
+        with self.assertRaises(TypeError):
+            empty_dset.write_direct(arr, np.s_[0:10], np.s_[50:60])
+
+        dset.write_direct(arr2, np.s_[0:10], np.s_[50:60])
+        self.assertEqual(dset.shape, (100,))
```

Comment:
Similarly here - it will pass, but it's not really relevant to what you're testing.